File: ps/leetcode/Add Two Numbers.cpp

struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
	public:
		ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
			ListNode* dummyNode = new ListNode(-1);
			ListNode* curNode = dummyNode;
			int remain = 0;
			while (l1 != NULL || l2 != NULL)
			{
				int n1 = l1!=NULL?l1->val:0;
				int n2 = l2!=NULL?l2->val:0;
				int num = n1 + n2 + remain;
				ListNode *newNode = new ListNode(num%10);
				remain = num / 10;
				curNode->next = newNode;
				curNode = newNode;
				if (l1 != NULL) l1 = l1->next;
				if (l2!= NULL) l2 = l2->next;
			}
			if (remain)
			{
				ListNode *newNode = new ListNode(1);
				curNode->next = newNode;
				curNode = newNode;
			}
			return dummyNode->next;
		}
};

File: ps/leetcode_java/src/AddTwoNumbers/AddTwoNumbers.java
package AddTwoNumbers;

public class AddTwoNumbers {
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode cur = dummy;
        int carry = 0;
        while(l1!=null || l2!=null){
            int v1 = l1==null ? 0 : l1.val;
            int v2 = l2==null ? 0 : l2.val;
            int val = (v1+v2+carry)%10;
            carry = (v1+v2+carry)/10;
            ListNode newNode = new ListNode(val);
            cur.next = newNode;
            cur = newNode;
            if(l1!=null) l1=l1.next;
            if(l2!=null) l2=l2.next;
        }
        if(carry==1){
            ListNode newNode = new ListNode(1);
            cur.next = newNode;
        }
        return dummy.next;
    }
}
